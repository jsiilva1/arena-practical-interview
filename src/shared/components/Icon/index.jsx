/**
 * External Dependencies
 */
import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { pascalCase } from 'change-case';
import * as FeatherIcons from 'react-feather';
import * as far from '@fortawesome/free-regular-svg-icons';
import * as fas from '@fortawesome/free-solid-svg-icons';
import * as fab from '@fortawesome/free-brands-svg-icons';

const faLib = {
  far,
  fas,
  fab,
};

/**
 * Icon Component
 *
 * @description Render a passed icon with props
 *
 * @param {Object} props component props.
 *
 * @return {*} component
*/
const Icon = (props) => {
  const {
    vendor = 'feather',
  } = props;

  let { name } = props;

  let result = '';

  const opts = { ...props };
  delete opts.vendor;

  opts.className = `${(opts.className || '')} arena-icon`;

  // Feather.
  if (vendor === 'feather') {
    name = pascalCase(name);

    if (!FeatherIcons[name]) {
      name = 'AlertTriangle';
    }

    const RenderIcon = FeatherIcons[name];

    result = <RenderIcon {...opts} strokeWidth="1.5" />;
  }

  // FontAwesome.
  if (vendor === 'fa') {
    let icon = '';
    let lib = 'far';

    if (Array.isArray(name)) {
      lib = name[0];
      name = name[1];
    }

    name = `fa${pascalCase(name)}`;

    if (faLib[lib] && faLib[lib][name]) {
      icon = faLib[lib][name];
    }

    result = icon ? <FontAwesomeIcon {...opts} icon={icon} /> : '';
  }

  return result;
};

export default Icon;
