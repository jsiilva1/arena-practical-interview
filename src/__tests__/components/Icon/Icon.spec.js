import React from 'react';
import { shallow } from 'enzyme';

import Icon from '../../../shared/components/Icon';

describe('Icon Component', () => {
  it('should render the Icon Component correctly', () => {
    const wrapper = shallow(
      <Icon
        name={['fas', 'check-circle']}
        vendor="fa"
      />
    );
    
    expect(wrapper).toMatchSnapshot();
  });
});
