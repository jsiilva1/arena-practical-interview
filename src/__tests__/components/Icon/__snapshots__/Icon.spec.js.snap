// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon Component should render the Icon Component correctly 1`] = `
<FontAwesomeIcon
  border={false}
  className=" arena-icon"
  fixedWidth={false}
  flip={null}
  icon={
    Object {
      "icon": Array [
        512,
        512,
        Array [],
        "f058",
        "M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z",
      ],
      "iconName": "check-circle",
      "prefix": "fas",
    }
  }
  inverse={false}
  listItem={false}
  mask={null}
  name={
    Array [
      "fas",
      "check-circle",
    ]
  }
  pull={null}
  pulse={false}
  rotation={null}
  size={null}
  spin={false}
  swapOpacity={false}
  symbol={false}
  title=""
  transform={null}
/>
`;
